In [1]:
import numpy as np
import pandas as pd

In [5]:
!pip install scikit-learn

In [12]:
import sklearn
print(sklearn.__version__)

0.24.2


In [10]:
## 필수라이브러리 호출

from sklearn.datasets import load_iris # 데이터 로드
from sklearn.model_selection import train_test_split ## X와 y분할
from sklearn.tree import DecisionTreeClassifier # 알고리즘
from sklearn.metrics import accuracy_score # 평가지표

import numpy as np
import pandas as pd

In [23]:
# load_iris()를 iris로 객체화

iris = load_iris()

X_ftrs = iris.data # X값들 = 특성(features) = 설명변수
y_target = iris.target # y값 = 목표변수 = 종속변수 = 반응변수
X_columns = iris.feature_names

In [27]:
# data.frame으로 변경
iris_df = pd.DataFrame(X_ftrs, columns=X_columns)
iris_df.head()
iris_df['label'] = y_target 
print(iris_df.head())
iris_df.tail()

   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   label  
0      0  
1      0  
2      0  
3      0  
4      0  


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     label  
145      2  
146      2  
147      2  
148      2  
149      2

In [29]:
# Train_data 및 Test_data 분할 :: Hold_out(홀드아웃)

다음으로 학습용 데이터와 테스트용 데이터를 분리해보자. 학습용 데이터와 테스트용 데이터는 반드시 분리해야 한다. 학습데이터로 학습된 모델이 얼마나 뛰어난 성능을 가지는지 평가하려면 테스트 데이터 세트가 필요하기 때문이다.  

In [136]:
# array1 = np.arange(1,11)
# array2 = np.arange(21,31)
# X_train, X_test, y_train, y_test = train_test_split(array1,array2, test_size=0.3, random_state=11)

In [152]:
### 데이터를 확인
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   label              150 non-null    int32  
dtypes: float64(4), int32(1)
memory usage: 5.4 KB


In [137]:
X_train, X_test, y_train, y_test = train_test_split(X_ftrs, y_target, test_size=0.2, random_state=11)

In [139]:
### 데이터의 크기 확인
print('X_train의 shape 확인:', X_train.shape)
print('X_test의 shape 확인:', X_test.shape)
print('y_train의 shape 확인:', y_train.shape)
print('y_test의 shape 확인:', y_test.shape)

X_train의 shape 확인: (120, 4)
X_test의 shape 확인: (30, 4)
y_train의 shape 확인: (120,)
y_test의 shape 확인: (30,)


In [140]:
### 학습알고리즘 호출 및 객체화
dt_clf = DecisionTreeClassifier(random_state=11)

In [141]:
# 학습을 수행
dt_clf.fit(X_train, y_train)

DecisionTreeClassifier(random_state=11)

In [144]:
X_test.shape

(30, 4)

In [147]:
# 학습된 dt_clf객체에서 테스트 데이터로 예측을 수행
pred_dt = dt_clf.predict(X_test)

In [149]:
## 모델을 평가
from sklearn.metrics import accuracy_score as acc_sc
result_acc = acc_sc(y_test,pred_dt)
print('예측 정확도:{0:.4f}'.format(result_acc))

예측 정확도:0.9333


학습한 의사 결정 트리의 알고리즘 예측 정확도가 약 0.9333(93.33%)으로 측정되었다. 앞의 붓꽃 데이터 세트로 분류를 예측한 프로세스를 정리하면 다음과 같다.

    1. 데이터 세트 분리: 데이터를 학습 데이터와 테스트 데이터로 분리한다.
    
    2. 모델 학습: 학습 데이터를 기반으로 ML 알고리즘을 적용해 모델을 학습시킨다.
    
    3. 예측 수행: 학습된 ML 모델을 이용해 테스트 데이터의 분류 (즉, 붓꽃 종류)를 예측한다.
    
    4. 평가: 이렇게 예측된 결과 값과 테스트 데이터의 실제 결과를 비교해 ML 모델 성능을 평가한다.

In [150]:
# end of file